### DISCUSSION TITLE
[Discussion]int <A,B,C> m =0;

### DISCUSSION THREAD

**Discussion by 6MiaoFly** [dotnet/csharplang#3523] [Language Ideas]
# 1.What I want to say :
Supposing the new syntax is named "Symbol".
It may be another type of "**Extension Methods**",but the syntax is for **instance** , not **class**.
It can be used in **declaration-statement**，or **run-time**.
And the type of "Symbol" can be **static** and **non-static**.
# 2.Declaration Statement
Usually，a declaration statement is :
```C#
int m = 0 ;
```
The the new syntax of declaration statement :
```C#
int <symbolA,symbolB,symbolC> m = 0 ; 
```
"symbolA,symbolB,symbolC" is all  "Symbol-Type" .Any of them is collection of functions for Type T,which include "int"."symbolA" have the function:MethodA1,MethodA2,MethodA3,...
Then , "m" can use method, like:
```C#
m.MethodA1();
m.MethodA2();
```
### 2.1 Static Symbol
"symbolA" can be a static class.Just one static instance for all instance of type T to use.
Like:
```C#
public static class SymbolA
{
　static int x = 0;
　public static void MethodA1(that object t)//keyword "this" is for Extension-Methods.
　　=> Console.WriteLine(t.ToString());
　public static void MethodA2()
　　=> x++;
　public static int MethodA3()
　　=> x++;
}
```
What is different with Extension-Methods ? see the code:
 ```C#
int <symbolA> a = 0 ; 
int b = 0 ; 
a.MethodA1();//It's OK.
b.MethodA1();//Error. "b" hasn't the method.
b=a;//OK , typeof(a)=typeof(b)=typeof(int)
b.MethodA1();//Error. "b" hasn't the method.
```
### 2.2 Non-tatic Symbol
"symbolB" can be a non-static class.One instance for one instance of type T to use.
Like:
```C#
using System.Symbol;//Maybe

public class symbolB<T>: Symbol<T> where T:struct
{
　int x=0;
　public void MethodB1(that object t)//Not this , this is for Extension-Methods.
　　=> Console.WriteLine(t.ToString());
　public void MethodB2()
　　=> x++;
　public int MethodB3()
　　=> x++;
}
```
In(1) ,there have two instances.One is "int" of a,the other is "symbolB\<int\>" of a.
 ```C#
int <symbolB> a = 0 ; //(1)
int <symbolB> b = 0 ; 
a.MethodB2();
int i=a.MethodB3();//i=1
int j=b.MethodB3();//j=0
```
Static-Symbol：
 ```C#
int <symbolA> a = 0 ; 
int <symbolA> b = 0 ; 
a.MethodA2();
int i=a.MethodA3();//i=1
int j=b.MethodA3();//j=2
```

# 3.Run-time Using
### 3.1 For the static Symbol:
SymbolD is a static Symbol.
Code maybe :
```C#
int a= 0 ;
<SymbolD>a.MethodD1();
```
Maybe :
```C#
int a= 0 ;
(SymbolD)a.MethodD1();
```
Maybe :
```C#
int a= 0 ;
(a as SymbolD).MethodD1();
```
### 3.2 For the non-static Symbol:
A SymbolB instance can use for many instances.SymbolB has been given.
Code maybe :
 ```C#
int a = 0 ; 
int b = 0 ; 
SymbolB<int> sym = new SymbolB<int>();
<sym>a.MethodB1();//Console.WriteLine(a.ToString());
<sym>a.MethodB2();//sym.x=1
sym.MethodB2();//sym.x=2
int i=<sym>a.MethodB3();//i=2,sym.x=3
i=<sym>b.MethodB3();//i=3,sym.x=4
i=sym.MethodB3();//i=4,sym.x=5
```
# 4.Combination Symbol

Code maybe:
 ```C#
(SymbolA, SymbolB)<int> sym= new <SymbolA<int>(), SymbolB<int>()>;
 ```
Maybe:
 ```C#
<SymbolA<int>, SymbolB<int>> sym= new <SymbolA<int>(), SymbolB<int>()>;
 ```
Maybe:
```C#
A a=new A();
(Symbol1, Symbol2)<B,C> sym= new <Symbol1<B>(), Symbol2<C>()>;
<sym>a.fun1();
//(a is B)=true
//(a is C)=true
```

# 5.Same Methods

For the same Algorithm-fun can have different Implement.
```C#
BaseClass <symbolAlgorithmA> g;
g.Method();
BaseClass <symbolAlgorithmB> h;
h.Method();
```
Then , code mabe: 
```C#
BaseClass <symbolAlgorithmA, symbolAlgorithmB> i;
i.symbolAlgorithmA.Method();
```
Maybe:
```C#
BaseClass <symbolAlgorithmA, symbolAlgorithmB> i;
<symbolAlgorithmA>i.Method();
```
# 6.Similar
Would #1695  be a similar approach ???

**Comment by svick** [comment]
One thing that I think is missing from this suggestion is answering "Why?": What is the point of all this extra syntax? How does it improve your code? What makes it better than the existing options?
